include/ginkgo/core/base/exception.hpp

```diff
@@ -269,6 +269,33 @@ class DimensionMismatch : public Error {
 };
 
 
+/**
+ * BadDimension is thrown if an operation is being applied to a LinOp of
+ * with bad dimensions.
+ */
+class BadDimension : public Error {
+public:
+    /**
+     * Initializes a bad dimension error.
+     * @param file The name of the offending source file
+     * @param line The source code line number where the error occurred
+     * @param func The function name where the error occurred
+     * @param first_name The name of the first operator
+     * @param first_rows The output dimension of the first operator
+     * @param first_cols The input dimension of the first operator
+     * @param clarification An additional message describing the error further
+     */
+    BadDimension(const std::string &file, int line, const std::string &func,
+                 const std::string &first_name, size_type first_rows,
+                 size_type first_cols, const std::string &clarification)
+        : Error(file, line,
+                func + ": bad dimension for " + first_name + " [" +
+                    std::to_string(first_rows) + " x " +
+                    std::to_string(first_cols) + "]: " + clarification)
```

Comment:
The error output is not very descriptive in my opinion.
I think it would read something like:
```
test_file.cpp:123: function_name: bad dimension for system_matrix [0, 2]: expected non-empty matrix
```
I would rewrite it, so it says something like
```
system_matrix has dimension [0, 2] while expecting: non-empty matrix
```

Also, I would rename `clarification` to `expected` if you change the text in that way.